feat(ExternalIds):Support multi-externalId Values of same externalId key in Project, Component, and Release page

albert-hjy opened the pull request:
[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.
> * Which issue is this pull request belonging to and how is it solving it? (https://github.com/eclipse/sw360/issues/865)
> * No dependencies updated

Issue: 

### Suggest Reviewer
> @mcjaeger 

### How To Test?
> How should these changes be tested by the reviewer?
**Project**
> 1. UI: Edit Project
![image](https://user-images.githubusercontent.com/61676785/85089719-ee486280-b215-11ea-98a3-c7112b66e1e2.png)
> 2. UI: Display Project
![image](https://user-images.githubusercontent.com/61676785/85089769-0a4c0400-b216-11ea-8dbe-16e054be21ce.png)
> 3. API: Search by external id
![image](https://user-images.githubusercontent.com/61676785/85089808-251e7880-b216-11ea-8887-924c7c38fb1a.png)

**Component**
> 1. UI: Edit Component
![image](https://user-images.githubusercontent.com/61676785/85089971-93633b00-b216-11ea-8d4e-e84166868ce5.png)

> 2. UI: Display Component
![image](https://user-images.githubusercontent.com/61676785/85089978-965e2b80-b216-11ea-80e0-4316f7dfb3e4.png)

> 3. API: Search by external id
![image](https://user-images.githubusercontent.com/61676785/85089987-9ceca300-b216-11ea-9fce-1a0c9770c56b.png)

**Release**
> 1. UI: Edit Release
![image](https://user-images.githubusercontent.com/61676785/85089992-9fe79380-b216-11ea-9cd7-22f155ab636e.png)

> 2. UI: Display Release
![image](https://user-images.githubusercontent.com/61676785/85090006-a70ea180-b216-11ea-9f46-9d3d22c8ed12.png)

> 3. API: Search by external id
![image](https://user-images.githubusercontent.com/61676785/85090016-aaa22880-b216-11ea-9218-3d33fa7daf4c.png)
![image](https://user-images.githubusercontent.com/61676785/85090023-aece4600-b216-11ea-8277-57355e9fd30a.png)

> No additional test

### Checklist
Must:
- [x] All related issues are referenced in commit messages and in PR

oheger-bosch commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortletUtils.java:
```diff
@@ -369,6 +371,27 @@ public static VerificationState getVerificationState(VulnerabilityDTO vul){
+                                            .findFirst()
+                                            .orElse("");
+                                }
+                                ObjectMapper mapper = new ObjectMapper();
```
Creating an object mapper is a rather expensive operation; so this should not be done for each element in the map. The object creation should at least be moved to the beginning of the method.

As the ObjectMapper class is thread-safe, in this case it would also be an option to have a static ObjectMapper field in PortletUtils that is always reused.

albert-hjy replied:
Thanks a lot, @oheger-bosch 
I have refactored the code.

oheger-bosch replied:
@albert-hjy, thanks for the update. This looks better now performance-wise. I am, however, a bit worried that the lazy initialization approach is not thread-safe. As the utils class is called during request execution, there will be multiple threads involved, and we should take care that this is safe.

The easiest solution is certainly to get rid off the lazy initialization, make the ObjectMapper field *final*, and create the mapper directly.

If you prefer lazy initialization, then I would suggest using the [Lazy initialization holder class idiom](https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom) for static fields.

Please also refer to J. Bloch, "Effective Java", 3rd edition, item 83 "Use lazy initialization judiciously"

albert-hjy replied:
@oheger-bosch , thanks a lot. Agree that it's unnecessary to use lazy initialization here. I have changed ObjectMapper field _final_.

albert-hjy commented:
Thank you @JaideepPalit , 
I have

- added `EscapeXml` to `DisplayMapOfExternalIdsString.java`
- added migration scripts for updating CouchDB Project, Component, and Release `byExternalId` view
- documented the two different types of value `String` and `Array of String` to Project, Component, and Release APIs
- fixed `update existing project or Component or Release without any change in External IDs. ExternalId value will be converted from String to Array of String`

JaideepPalit commented on scripts/migrations/020_update_byExternalIds_release_view.py:
```diff
@@ -0,0 +1,97 @@
+
+import time
+import couchdb
+import jsonß
```
Getting error while executing this script

```
File "020_update_byExternalIds_release_view.py", line 21
SyntaxError: Non-ASCII character '\xc3' in file 020_update_byExternalIds_release_view.py on line 21, but no encoding declared; see http://python.org/dev/peps/pep-0263/ for details
```

albert-hjy commented:
Thank you @JaideepPalit 
- There are some error occurring while running migration script scripts/migrations/020_update_byExternalIds_release_view.py due to presence of foreign character
> This was a typo when I change the CouchDB from our remote testing server to localhost. Fixed it.
- Kindly provide same read/write/execute access for newly added migration script
> Added the same permission to new migration scripts.
![image](https://user-images.githubusercontent.com/61676785/87943639-e3ad1180-cad0-11ea-8ba1-201c9d4ec3a4.png)

mcjaeger commented:
hi, just tested it on updating the PR over an installation which had current master :
* install ok
* multiple values for same id works and stored and displayed properly

However, executing the migration scripts brings (from migration log):
```
{
    "old_value": "function(doc) {  if (doc.type == 'component') {    for (var externalId in doc.externalIds) {      try {            var values = JSON.parse(doc.externalIds[externalId]);            for (
var idx in values) {              emit( [externalId, values[idx]], doc._id);            }      } catch(error) {          emit( [externalId, doc.externalIds[externalId]], doc._id);      }    }  }}", 
    "field_to_update": "views->byExternalIds->map", 
    "docId": "_design/Component", 
    "new_value": "function(doc) {  if (doc.type == 'component') {    for (var externalId in doc.externalIds) {      try {            var values = JSON.parse(doc.externalIds[externalId]);            for (
var idx in values) {              emit( [externalId, values[idx]], doc._id);            }      } catch(error) {          emit( [externalId, doc.externalIds[externalId]], doc._id);      }    }  }}"
}
```

-> I am not sure if that is serious, because the records had external ids before and everything seems to work ...

mcjaeger commented:
wow, even search works for newly added multiple values from the same key ...

mcjaeger commented:
@JaideepPalit I think it works. if it looks OK to you, could you pls. approve and merge?
